### PR TITLE
Implement a field selector for ClusterIP on Services

### LINKED ISSUE #122394: Support skip watch Headless Service in server-side.
### What would you like to be added?

Support watch service with field-selector spec.clusterIP!=None to skip headless service in kube-apiserver. And kubelet/kube-proxy can skip watch headless-service in the future.
image

Now, endpoints will be added labels [service.kubernetes.io/headless](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpoint/endpoints_controller.go#L518) for headless service  by endpoints-controller,  and kube-proxy will ignore  event for it by label-selector. https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-proxy/app/server.go#L913. But headless service haven't be added label and kube-proxy  still watch for it.

For kubelet, it watch service to add service-env for pods , but it ignores headless service when  build service-env-map(https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_pods.go#L549). So kubelet don't need to watch headless service either.

### Why is this needed?

In a large-cluster with more then 5000 nodes,  when we run kubeflow task( tfjob:pod:headlessService=1:10:10) with qps 10, it can save huge network bandwidth and cpu for kube-apiserver.

### PR DESCRIPTION
**Pull request by aojea**
#### What type of PR is this?

/kind feature

Implement a field selector for services that allows users to filter by the clusterIP field.
Use this new field to filter server side headless Services on Kubelet to improve the scalability on environments with large number of Headless Services

Fixes #122394

```release-note
NONE
```

Reverted in https://github.com/kubernetes/kubernetes/pull/123896/

### CODE REVIEW

**Comment by Sakuralbj** [comment]
Submit a similar PR [https://github.com/kubernetes/kubernetes/pull/122543](url) for this,  add an additional change in addConversionFuncs, and support indexField for spec.clusterIP as well. If the change can be added in this pr together

**Comment by aojea** [comment]
> Submit a similar PR [https://github.com/kubernetes/kubernetes/pull/122543](url) for this, add an additional change in addConversionFuncs, and support indexField for spec.clusterIP as well. If the change can be added in this pr together

let me check , CI is failing so most probably I will need to add those changes,

File: cmd/kube-proxy/app/server.go [hunk 1]
```diff
@@ -904,11 +904,21 @@ func (s *ProxyServer) Run() error {
...
 	labelSelector := labels.NewSelector()
 	labelSelector = labelSelector.Add(*noProxyName, *noHeadlessEndpoints)
 
-	// Make informers that filter out objects that want a non-default service proxy.
-	informerFactory := informers.NewSharedInformerFactoryWithOptions(s.Client, s.Config.ConfigSyncPeriod.Duration,
+	// Create informer for Services, filter out the ones that want a non-default service proxy or a Headless Service.
+	serviceInformer := informers.NewSharedInformerFactoryWithOptions(s.Client, s.Config.ConfigSyncPeriod.Duration,
+		informers.WithTweakListOptions(func(options *metav1.ListOptions) {
+			options.LabelSelector = labelSelector.String()
+			options.FieldSelector = fields.OneTermNotEqualSelector("spec.clusterIP", v1.ClusterIPNone).String()
```
Comment (aojea) [inline cmd/kube-proxy/app/server.go:917]: this will fail against an old apiserver :/

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -451,7 +451,11 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
 	var serviceLister corelisters.ServiceLister
 	var serviceHasSynced cache.InformerSynced
 	if kubeDeps.KubeClient != nil {
-		kubeInformers := informers.NewSharedInformerFactoryWithOptions(kubeDeps.KubeClient, 0)
+		// don't watch headless services, they are not needed since this informer is only used to create the environment variables for pods.
+		// See https://issues.k8s.io/122394
+		kubeInformers := informers.NewSharedInformerFactoryWithOptions(kubeDeps.KubeClient, 0, informers.WithTweakListOptions(func(options *metav1.ListOptions) {
+			options.FieldSelector = fields.OneTermNotEqualSelector("spec.clusterIP", v1.ClusterIPNone).String()
```
Comment (aojea) [inline pkg/kubelet/kubelet.go:457]: this will fail against an old apiserver :/

Comment (aojea) [reply pkg/kubelet/kubelet.go:457]: this is covered on our skew policies though https://kubernetes.io/releases/version-skew-policy/

File: cmd/kube-proxy/app/server.go (hunk 1 above)
Comment (aojea) [reply cmd/kube-proxy/app/server.go:917]: this is covered on our skew policies though https://kubernetes.io/releases/version-skew-policy/

File: pkg/apis/core/v1/conversion.go [hunk 3]
```diff
@@ -97,6 +97,9 @@ func addConversionFuncs(scheme *runtime.Scheme) error {
 	if err := AddFieldLabelConversionsForSecret(scheme); err != nil {
 		return err
 	}
+	if err := AddFieldLabelConversionsForService(scheme); err != nil {
```
Comment (aojea) [inline pkg/apis/core/v1/conversion.go:100]: @Sakuralbj I added you as co-author of this commit

File: test/integration/service/service_test.go [hunk 4]
```diff
@@ -264,3 +269,53 @@ func Test_RemovingExternalIPsFromClusterIPServiceDropsExternalTrafficPolicy(t *t
...
+	}))
+	serviceLister := kubeInformers.Core().V1().Services().Lister()
+	serviceHasSynced := kubeInformers.Core().V1().Services().Informer().HasSynced
+	kubeInformers.Start(ctx.Done())
+	cache.WaitForCacheSync(ctx.Done(), serviceHasSynced)
+	svcs, err := serviceLister.List(labels.Everything())
+	if err != nil {
+		t.Fatalf("Error listing services: %v", err)
+	}
+	// onlh the kubernetes.default service expected
```
Comment (HirazawaUi) [inline test/integration/service/service_test.go:317]:
```suggestion
	// only the kubernetes.default service expected
```

File: pkg/registry/core/service/strategy.go [hunk 5]
```diff
@@ -162,6 +167,33 @@ func (serviceStatusStrategy) WarningsOnUpdate(ctx context.Context, obj, old runt
 	return nil
 }
 
+// GetAttrs returns labels and fields of a given object for filtering purposes.
+func GetAttrs(obj runtime.Object) (labels.Set, fields.Set, error) {
+	service, ok := obj.(*api.Service)
+	if !ok {
+		return nil, nil, fmt.Errorf("not a service")
+	}
+	return labels.Set(service.Labels), SelectableFields(service), nil
```
Comment (HirazawaUi) [inline pkg/registry/core/service/strategy.go:176]:
```suggestion
	return service.Labels, SelectableFields(service), nil
```
Unnecessary type conversion.

**Comment by Sakuralbj** [comment]
Is this change only considered for support in newer versions of Kubernetes?

**Comment by Sakuralbj** [comment]
Is this change only considered for support in newer versions of Kubernetes?

**Comment by aojea** [comment]
> Is this change only considered for support in newer versions of Kubernetes?

yes, this is a new feature and is not meeting the requirements for cherry-picks

**Comment by Sakuralbj** [comment]
> > Is this change only considered for support in newer versions of Kubernetes?
> 
> yes, this is a new feature and is not meeting the requirements for cherry-picks

Thanks for your reply.  BTW,  what's the problem with the current PR being merged?  I wonder if we can use this optimization in  k8s 1.30 ?

**Comment by aojea** [comment]
> Thanks for your reply. BTW, what's the problem with the current PR being merged?

we need reviewers and approvers , and some people is on vacation

@sttts @liggitt can you PTAL to the field selector bits?

File: pkg/apis/core/v1/conversion.go [hunk 6]
```diff
@@ -488,6 +491,21 @@ func AddFieldLabelConversionsForSecret(scheme *runtime.Scheme) error {
 		})
 }
 
+func AddFieldLabelConversionsForService(scheme *runtime.Scheme) error {
+	return scheme.AddFieldLabelConversionFunc(SchemeGroupVersion.WithKind("Service"),
+		func(label, value string) (string, string, error) {
+			switch label {
+			case "type",
```
Comment (liggitt) [inline pkg/apis/core/v1/conversion.go:498]: `"type"`?

**Review by liggitt — COMMENTED** [review]
this looks (mostly) mechnically correct

one copy/paste typo and one question about whether event handlers are correctly handling the Add / Delete events that will result from this if the spec.clusterIP field changes from / to None

**Comment by Sakuralbj** [comment]
> Remove watching Headless Services for kubelet/kube-proxy #122502



> this looks (mostly) mechnically correct
> 
> one copy/paste typo and one question about whether event handlers are correctly handling the Add / Delete events that will result from this if the spec.clusterIP field changes from / to None

Thanks for your review. I do a test and find spec.clusterIP can not change once set.  We will receive an error when edit service spec.clusterIP.
image
related validation:
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L5620
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L7695

**Comment by liggitt** [comment]
I'm pretty sure there are transitions that allow setting / clearing clusterIP

for example:

```yaml
kind: Service
apiVersion: v1
metadata:
  name: test
spec:
  type: ExternalName
  externalName: foo
```

to 

```yaml
kind: Service
apiVersion: v1
metadata:
  name: test
spec:
  clusterIP: None
```

and back

File: pkg/apis/core/v1/conversion.go (hunk 6 above)
Comment (aojea) [reply pkg/apis/core/v1/conversion.go:498]: copy paste :(

File: cmd/kube-proxy/app/server.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (aojea) [reply cmd/kube-proxy/app/server.go:925]: added integration test to validate it works correctly

File: test/integration/service/service_test.go [hunk 7]
```diff
@@ -264,3 +272,140 @@ func Test_RemovingExternalIPsFromClusterIPServiceDropsExternalTrafficPolicy(t *t
...
+				oldSvc := oldObj.(*v1.Service)
+				newSvc := newObj.(*v1.Service)
+				t.Logf("Updated Service %#v to %#v", oldSvc, newSvc)
+			},
+			DeleteFunc: func(obj interface{}) {
+				svc := obj.(*v1.Service)
+				t.Logf("Deleted Service %#v", svc)
+			},
+		},
+	); err != nil {
```
Comment (aojea) [inline test/integration/service/service_test.go:339]: @liggitt I prefer to keep these handlers for debugging, those were very useful to me.
 I think is enough validating the state of the cache as the test is doing now, Do you want me to validate the events transitions too?

**Comment by Sakuralbj** [comment]
> transitions



> I'm pretty sure there are transitions that allow setting / clearing clusterIP
> 
> for example:
> 
> ```yaml
> kind: Service
> apiVersion: v1
> metadata:
>   name: test
> spec:
>   type: ExternalName
>   externalName: foo
> ```
> 
> to
> 
> ```yaml
> kind: Service
> apiVersion: v1
> metadata:
>   name: test
> spec:
>   clusterIP: None
> ```
> 
> and back


Thanks, the transitions actually can success，I missed test for ExternalName service. I see the check for validate-service-update will skip clusterIP-validate when update service from or to ExternalName service.  So there is possibility  headless-svc change to ExternalName  or ExternalName to headless.
image
image

**Comment by aojea** [comment]
> Thanks, the transitions actually can success

that was not on purpose, for the records, is a side effect of how complex is the Service API

**Comment by Sakuralbj** [comment]
Hi [@aojea](https://github.com/aojea) , is there any progress on this?  
Our ML-training cluster can save huge traffic with this commit and we are looking forward to the update，if we can pass test first to  move it forward，thanks!

File: pkg/registry/core/service/strategy.go [hunk 8]
```diff
@@ -162,6 +167,33 @@ func (serviceStatusStrategy) WarningsOnUpdate(ctx context.Context, obj, old runt
...
+func GetAttrs(obj runtime.Object) (labels.Set, fields.Set, error) {
+	service, ok := obj.(*api.Service)
+	if !ok {
+		return nil, nil, fmt.Errorf("not a service")
+	}
+	return service.Labels, SelectableFields(service), nil
+}
+
+// Matcher returns a selection predicate for a given label and field selector.
+func Matcher(label labels.Selector, field fields.Selector) pkgstorage.SelectionPredicate {
```
Comment (thockin) [inline pkg/registry/core/service/strategy.go:180]: Only used in the test - define it there?

File: pkg/registry/core/service/strategy_test.go [hunk 9]
```diff
@@ -782,3 +784,79 @@ func TestDropTypeDependentFields(t *testing.T) {
...
+			in: &api.Service{
+				Spec: api.ServiceSpec{ClusterIP: api.ClusterIPNone},
+			},
+			fieldSelector: fields.ParseSelectorOrDie("spec.clusterIP=2001:db2::1"),
+			expectMatch:   false,
+		},
+		{
+			name:          "no match on empty service",
+			in:            &api.Service{},
+			fieldSelector: fields.ParseSelectorOrDie("spec.clusterIP=None"),
```
Comment (thockin) [inline pkg/registry/core/service/strategy_test.go:846]: Add cases for `spec.clusterIP!=None` and maybe `spec.clusterIP!=""` (not sure if that last actually works!!)

File: test/integration/service/service_test.go [hunk 10]
```diff
@@ -264,3 +271,140 @@ func Test_RemovingExternalIPsFromClusterIPServiceDropsExternalTrafficPolicy(t *t
...
+			},
+		},
+	}
+
+	_, err = client.CoreV1().Services(ns.Name).Create(ctx, service, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Error creating test service: %v", err)
+	}
+
+	// informer to watch only headless services
```
Comment (thockin) [inline test/integration/service/service_test.go:314]: Comment says "watch headless" but code say "not equal to None"

**Review by thockin — COMMENTED** [review]
Should we also select on Type ?

File: pkg/registry/core/service/strategy.go (hunk 8 above)
Comment (aojea) [reply pkg/registry/core/service/strategy.go:180]: copied from the other places where this is implemented

```
pkg/registry/core/secret/storage/storage.go:            PredicateFunc:             secret.Matcher,
pkg/registry/core/secret/strategy.go:// Matcher returns a selection predicate for a given label and field selector.
pkg/registry/core/secret/strategy.go:func Matcher(label labels.Selector, field fields.Selector) pkgstorage.SelectionPredicate {
pkg/registry/core/service/strategy_test.go:                     m := Matcher(labels.Everything(), testCase.fieldSelector)
pkg/registry/core/service/strategy.go:// Matcher returns a selection predicate for a given label and field selector.
pkg/registry/core/service/strategy.go:func Matcher(label labels.Selector, field fields.Selector) pkgstorage.SelectionPredicate {
pkg/registry/core/event/storage/storage.go:             PredicateFunc: event.Matcher,
pkg/registry/core/event/strategy.go:// Matcher returns a selection predicate for a given label and field selector.
pkg/registry/core/event/strategy.go:func Matcher(label labels.Selector, field fields.Selector) storage.SelectionPredicate {
pkg/registry/core/configmap/storage/storage.go:         PredicateFunc:             configmap.Matcher,
pkg/registry/core/configmap/strategy.go:// Matcher returns a selection predicate for a given label and field selector.
```

File: pkg/registry/core/service/storage/storage.go [hunk 11]
```diff
@@ -99,7 +99,10 @@ func NewREST(
 
 		TableConvertor: printerstorage.TableConvertor{TableGenerator: printers.NewTableGenerator().With(printersinternal.AddHandlers)},
 	}
-	options := &generic.StoreOptions{RESTOptions: optsGetter}
+	options := &generic.StoreOptions{
```
Comment (wojtek-t) [inline pkg/registry/core/service/storage/storage.go:102]: Set PredicateFunc when creating Store above.

File: pkg/registry/core/service/strategy.go (hunk 8 above)
Comment (wojtek-t) [reply pkg/registry/core/service/strategy.go:180]: And you missed to set the PredicateFunc when creating the store - added a comment above.

File: test/integration/service/service_test.go [hunk 12]
```diff
@@ -264,3 +271,140 @@ func Test_RemovingExternalIPsFromClusterIPServiceDropsExternalTrafficPolicy(t *t
...
+		if svc == nil || err != nil {
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		t.Fatalf("Error waiting for test service test-clusterip: %v", err)
+	}
+
+	// mutate the Service to drop the ClusterIP, theoretically ClusterIP is inmutable but ...
```
Comment (wojtek-t) [inline test/integration/service/service_test.go:372]: Hmm - this seems hacky... Why not just create a headless service?

Comment (aojea) [reply test/integration/service/service_test.go:372]: this is the case Jordan brought in the reviews https://github.com/kubernetes/kubernetes/pull/122541#pullrequestreview-1828839818

File: pkg/registry/core/service/strategy.go (hunk 8 above)
Comment (aojea) [reply pkg/registry/core/service/strategy.go:180]: done

File: test/integration/service/service_test.go (hunk 12 above)
Comment (aojea) [reply test/integration/service/service_test.go:372]: the test it is actually just to verify that corner case does not break controllers

**Comment by thockin** [comment]
Does this warrant a gate?  I don't THINK so, but need to ask.

Should we also support selecting on .Type?  Now that scaffolding is in place, adding more would be easy.

/lgtm
/approve
/hold   

Remove hold if you are OK with the above.

**Comment by aojea** [comment]
> Does this warrant a gate? I don't THINK so, but need to ask.


> Should we also support selecting on .Type? Now that scaffolding is in place, adding more would be easy.

for 1.31 :) 

> 
> /lgtm /approve /hold

/hold cancel

**Comment by sftim** [comment]
No release note? Usually we add those when changing behavior.

**Comment by aojea** [comment]
> No release note? Usually we add those when changing behavior.

it was reverted @sftim https://github.com/kubernetes/kubernetes/pull/123896

**Pull request merged**